Backend/Remora.Discord.Voice/DiscordVoiceClient.cs

```diff
@@ -0,0 +1,1258 @@
+        public async Task<Result> TransmitAudioAsync
+        (
+            Stream audioStream,
+            SpeakingFlags speakingFlags = SpeakingFlags.Microphone,
+            IAudioTranscoderService? transcoder = null,
+            CancellationToken ct = default
+        )
+        {
+            if (this.ConnectionStatus is not GatewayConnectionStatus.Connected || !_dataService.IsConnected)
+            {
+                return new InvalidOperationError("Gateway is not connected.");
+            }
+
+            if (speakingFlags is not SpeakingFlags.Microphone or SpeakingFlags.Priority)
+            {
+                return new InvalidOperationError($"Speaking flags must not be set to {nameof(SpeakingFlags.Microphone)} or {nameof(SpeakingFlags.Priority)}.");
+            }
+
+            if (transcoder is null)
+            {
+                transcoder = _defaultTranscoder;
+                await transcoder.ResetAsync(ct).ConfigureAwait(false);
+            }
+
+            if (!transcoder.IsInitialized)
+            {
+                return new InvalidOperationError("The provided transcoder must first be initialized.");
+            }
+
+            // Take a local copy of the SSRC so we can still send a voice state update when the client is stopped
+            var ssrc = _voiceServerConnectionDetails!.SSRC;
+            var needsRelease = false;
+
+            var sampleSize = transcoder.SampleSize;
+            var inputBuffer = MemoryPool<byte>.Shared.Rent(sampleSize);
+            var opusBuffer = MemoryPool<byte>.Shared.Rent(sampleSize);
+
+            try
+            {
+                if (!await _transmitSemaphore.WaitAsync(10, ct).ConfigureAwait(false))
+                {
+                    return new InvalidOperationError("This client is already transmitting audio.");
+                }
+
+                needsRelease = true;
+
+                // From https://github.com/DSharpPlus/DSharpPlus/blob/master/DSharpPlus.VoiceNext/VoiceNextConnection.cs
+                var synchronizerTicks = (double)Stopwatch.GetTimestamp();
+                var synchronizerResolution = Stopwatch.Frequency * 0.005;
+                var tickResolution = 10_000_000.0 / Stopwatch.Frequency;
+
+                EnqueueCommand
+                (
+                    new VoiceSpeakingCommand
+                    (
+                        speakingFlags,
+                        0,
+                        ssrc
+                    )
+                );
+
+                while (!ct.IsCancellationRequested)
+                {
+                    var read = await audioStream.ReadAsync
+                    (
+                        inputBuffer.Memory[0..sampleSize],
+                        ct
+                    ).ConfigureAwait(false);
+
+                    if (read != sampleSize)
+                    {
+                        break; // TODO: Does this cut off audio in some cases?
```

Comment:
To answer the question in this TODO, I'm no expert in audio, but my answer is it depends?

If you're getting the audio from something like a remote server, if we're read to the end of the stream, but the chunk doesn't align with a sample size (which, not sure how likely that is to happen), then there's suddenly no more audio because the server didn't split audio chunks in an apt way.

I *think* the better solution (mind you, I'm an idiot to take this with a bottle of salt) would to be check if the read is 0, which indicates that the stream is empty, and probably finished, though I suppose the cancelation token could also serve that purpose, if an end user is monitoring how long they're streaming audio.

This also brings up the issue of pausing, because a stream might get cut short, but the user might want to continue at a later point in time, which is generally fine, but will also require re-calling this method, so ehhh maybe I'm just being pessimistic.

Reply:
It's a good point. You have to rely on the stream being 'whole' before dispatching it. I'm considering re-writing this using `System.IO.Pipelines`; it would make situations like this easier to manage as the caller can provide data throughout the process, and the callee can read its desired chunks, waiting until the pipe is closed.

Another thing I'll play around with is filling out the remainder of a sample with zeroes, if the amount of bytes read from the stream is less than expected. This would let me utilise your idea of finishing when zero bytes are read; I naively terminate as I do currently because Opus expects a full sample.

I think pausing should be managed by the caller. There's no issue in re-calling `Transmit..` that I can see. Definitely open to discussion though.